primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1001 @@
+			}
+		}
+		Ok((Self::from_seed(&acc), Some(acc)))
+	}
+
+	/// Get the public key.
+	fn public(&self) -> Public {
+		let public = self.0.to_public();
+		let mut raw = [0; PUBLIC_SERIALIZED_LEN];
+		public
+			.serialize_compressed(raw.as_mut_slice())
+			.expect("key buffer length is good; qed");
+		Public::unchecked_from(raw)
+	}
+
+	/// Sign raw data.
+	fn sign(&self, data: &[u8]) -> Signature {
+		let data = vrf::VrfSignData::new(SIGNING_CTX, &[data], vrf::VrfIosVec::default());
+		self.vrf_sign(&data).signature
+	}
+
+	/// Verify a signature on a message.
+	///
+	/// Returns `true` if the signature is good.
+	fn verify<M: AsRef<[u8]>>(signature: &Self::Signature, data: M, public: &Self::Public) -> bool {
+		let data = vrf::VrfSignData::new(SIGNING_CTX, &[data.as_ref()], vrf::VrfIosVec::default());
+		let signature =
+			vrf::VrfSignature { signature: *signature, vrf_outputs: vrf::VrfIosVec::default() };
+		public.vrf_verify(&data, &signature)
+	}
+
+	/// Return a vector filled with seed raw data.
+	fn to_raw_vec(&self) -> Vec<u8> {
+		// TODO @burdges: we need a serializable Secret in dleq_vrf?
+		unimplemented!()
+	}
+}
+
+#[cfg(feature = "full_crypto")]
+impl CryptoType for Pair {
+	type Pair = Pair;
+}
+
+/// Bandersnatch VRF types and operations.
+pub mod vrf {
+	use super::*;
+	use crate::{bounded::BoundedVec, crypto::VrfCrypto, ConstU32};
+	use bandersnatch_vrfs::{
+		CanonicalDeserialize, CanonicalSerialize, IntoVrfInput, Message, PublicKey,
+		ThinVrfSignature, Transcript,
+	};
+
+	/// Max number of inputs/outputs which can be handled by the VRF signing procedures.
+	pub const MAX_VRF_IOS: u32 = 3;
+
+	/// Bounded vector used for VRF inputs and outputs.
+	///
+	/// Can contain at most [`MAX_VRF_IOS`] elements.
+	pub type VrfIosVec<T> = BoundedVec<T, ConstU32<MAX_VRF_IOS>>;
+
+	/// VRF input to construct a [`VrfOutput`] instance and embeddable within [`VrfSignData`].
+	#[derive(Clone, Debug)]
+	pub struct VrfInput(pub(super) bandersnatch_vrfs::VrfInput);
+
+	impl VrfInput {
+		/// Build a new VRF input.
+		///
+		/// Each message tuple has the form: message_data := (sub-domain, data).
+		pub fn new(domain: &'static [u8], message_data: &[(&[u8], &[u8])]) -> Self {
+			// ⚠️ TODO @davxy @burdges (temporary hack and probably needs to be fixed)
+			// In sassafras we want to construct a single `VrfInput` from multiple datas
+			// E.g. the ticket score uses: epoch-randomness, attempt-index, epoch-index
+			// Currently, `bandersnatch_vrfs::Message` has a single (domain, data) fields,
+			// so we need a workaround here...
+			let mut buf = Vec::new();
+			message_data.into_iter().for_each(|(sub_domain, data)| {
+				buf.extend_from_slice(sub_domain);
+				buf.extend_from_slice(&sub_domain.len().to_le_bytes());
+				buf.extend_from_slice(data);
+				buf.extend_from_slice(&data.len().to_le_bytes());
+			});
+			let msg = Message { domain, message: buf.as_slice() };
+			VrfInput(msg.into_vrf_input())
```

Comment:
@burdges this is the thing requiring attention.

As per comment. Sassafras ticket score VrfInput is constructed using 3 things: epoch-randomness, attempt-index, epoch-index (as also the pseudo code you've provided is doing [here](https://github.com/w3f/ring-vrf/blob/54f814faf33d5474d4901f5c9c965aec67a11f85/sassafras/src/tickets.rs#L73-L78)).

So this `VrfInput::new()` will be called with something like:

```
let in = VrfInput::new(
    "sassafras-ticket",
     &[
         (b"randomness", epoch_randomness_bytes),
         (b"attempt-index", attempt_index.to_le_bytes()),
         (b"epoch-index", epoch_index.to_le_bytes()),
     ],
 );
```

Because `bandersnatch_vrf::Message` is composed by a single domain and single data buffer, here I've used this hack to get sub-domains separation for the single fields.

If you're going to change it in the backend then I'll remove this

Reply:
We're just trying to fit into the IRTF H2C pattern here, so we'll make their domain thing meaningful.  I'd say something like this:
```
let domain = b"Sassafras v1"; 
let mut message = Vec::new();
{
    let f = |bytes: &[u8]| {
        message.extend_from_slice(bytes);
        let len = u32::try_from(bytes.len()).expect("No way we hash this much");
        message.extend_from_slice(len.to_le_bytes());
    }
    f(epoch_randomness_bytes);
    f(attempt_index.to_le_bytes());
    f(epoch_index.to_le_bytes());
}
let message = message.as_slice();
```
We could skip either the version number or the lengths, but maybe we just keep both I guess.